src/Routing/PushState.purs

```diff
@@ -0,0 +1,247 @@
+
+  -- The hashchange interface is asynchronous, since hashchange events are
+  -- fired on the next tick of the event loop. We want the push-state
+  -- interface to behave as similarly as possible, so we use the microtask
+  -- queue via MutationObserver to schedule callbacks. Alternatively we could
+  -- just use a setTimeout, but it would not be as prompt. We use a fresh
+  -- counter so that the text change mutation always fires.
+  schedule <- do
+    obsvNode <-
+      DOM.window
+        >>= DOM.document
+        >>> map DOM.htmlDocumentToDocument
+        >>= DOM.createTextNode ""
+        >>> map DOM.textToNode
+    observer <- DOM.mutationObserver \_ _ -> do
+      argsQueue <- readRef argsRef
+      writeRef argsRef []
+      for_ argsQueue notify
+    DOM.observe obsvNode { characterData: true } observer
+    pure \args -> do
+      argsQueue <- readRef argsRef
+      writeRef argsRef (Array.snoc argsQueue args)
+      when (Array.null argsQueue) do
+        fresh ← readRef freshRef
+        writeRef freshRef (fresh + 1)
+        DOM.setNodeValue (show fresh) obsvNode
+
+  let
+    stateFn op state path = do
+      let
+        loc = normalizeLocation state path
+        url = DOM.URL $ unwrap basename <> loc.pathname <> loc.search <> loc.hash
+      DOM.window
+        >>= DOM.history
+        >>= op state (DOM.DocumentTitle "") url
+      schedule loc
+
+    listener =
+      DOM.eventListener \_ -> notify =<< locationState
+
+  DOM.window
+    >>= DOM.windowToEventTarget
+    >>> DOM.addEventListener DOM.popstate listener false
+
+  pure
+    { pushState: stateFn DOM.pushState
+    , replaceState: stateFn DOM.replaceState
+    , locationState
+    , listen
+    }
+
+type LocationState =
+  { state :: Foreign
+  , path :: String
+  , pathname :: String
+  , search :: String
+  , hash :: String
+  }
+
+normalizeLocation :: Foreign -> Path -> LocationState
+normalizeLocation state (Path path) =
+  case searchIx, hashIx of
+    Nothing, Nothing   -> { state, pathname: path, hash: "", search: "", path }
+    Nothing, Just hix  -> { state, pathname: String.take hix path, search: "", hash: String.drop hix path, path }
+    Just six, Nothing  -> { state, pathname: String.take six path, search: String.drop six path, hash: "", path }
+    Just six, Just hix -> { state, pathname: String.take six path, search: String.take (hix - six) (String.drop six path), hash: String.drop hix path, path }
+  where
+  searchIx = String.indexOf (String.Pattern "?") path
+  hashIx = String.indexOf (String.Pattern "#") path
+
+-- | Folds effectfully over location changes given callbacks for handling
+-- | changes and the initial location. Returns an effect which removes the
+-- | listener.
+foldLocations
+  :: forall eff a
+   . (a -> LocationState -> Eff (PushStateEffects eff) a)
+  -> (LocationState -> Eff (PushStateEffects eff) a)
+  -> PushStateInterface (PushStateEffects eff)
+  -> Eff (PushStateEffects eff) (Eff (PushStateEffects eff) Unit)
```

Comment:
:+1: for effect removing listener. Could you wrap it into newtype though? `Eff _ Unit` looks a bit enigmatic in this case :)

Reply:
Routing.Hashes does not use a newtype. Maybe we can save that for the next breaking change?

Reply:
Okay